cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/Roller.java

```diff
@@ -0,0 +1,198 @@
+
+    /**
+     * Returns a future which completes when the given pod is "ready" after being rolled
+     * (and thus the next pods can be rolled, or the rolling restart is completed).
+     */
+    protected abstract Future<Void> afterRestart(Pod pod);
+
+    /**
+     * Returns a Future which completed with the actual pod corresponding to the abstract representation
+     * of the given {@code pod}.
+     */
+    protected abstract Future<Pod> pod(StatefulSet ss, P pod);
+
+    private static String getPodUid(Pod resource) {
+        if (resource == null || resource.getMetadata() == null) {
+            return NO_UID;
+        }
+        return resource.getMetadata().getUid();
+    }
+
+    /**
+     * Perform a rolling restart of the pods in the given StatefulSet.
+     * Pods will be tested for whether the really need rolling using the given {@code podRestart}.
+     * If a pod does indeed need restarting {@link #beforeRestart(Pod)} is called before and
+     * {@link #afterRestart(Pod)} is called afterwards.
+     * The returned Future is completed when the rolling restart is completed.
+     */
+    public Future<Void> rollingRestart(StatefulSet ss,
+                                Secret clusterCaCertSecret, Secret coKeySecret, Predicate<Pod> podRestart) {
+        return context(ss, clusterCaCertSecret, coKeySecret).compose(new Function<C, Future<C>>() {
+                @Override
+                public Future<C> apply(C context) {
+                    return sort(context, podRestart).recover(error -> {
+                        log.warn("Error when determining next pod to roll", error);
+                        return Future.succeededFuture(context);
+                    }).compose(currentContext -> {
+                        P pod = context.next();
+                        if (pod == null) {
+                            return Future.succeededFuture();
+                        } else {
+                            return pod(ss, pod).compose(p -> {
+                                log.debug("Rolling pod {} (still to consider: {})", p.getMetadata().getName(), currentContext);
+                                Future<Void> f = restartWithCallbacks(ss, p);
+                                return f.map(i -> currentContext).compose(this);
+                            });
+                        }
+                    });
+                }
+            }).map((Void) null);
+    }
+
+
+    /**
+     * Asynchronously apply the before restart callback, then restart the given pod
+     * by deleting it and letting it be recreated by K8s, then apply the after restart callback.
+     * Return a Future which completes when the after restart callback for the given pod has completed.
+     * @param ss The StatefulSet.
+     * @param pod The Pod to restart.
+     * @return a Future which completes when the after restart callback for the given pod has completed.
+     */
+    private Future<Void> restartWithCallbacks(StatefulSet ss, Pod pod) {
```

Comment:
maybe this one is `restart` and the other one `doPodRestart` or `doRestart` ... it's just that this `WithCallbacks` sounds strange to me :-)

Reply:
I think withCallbacks is more explicit. Using `restart` and `doRestart` it would not be clear how they differed without looking at the code. But we should tighten up the terminology. Are they callbacks or conditions? I suppose they're more like barriers.

Reply:
For sure "callbacks" is too much generic, is it too long `restartWithPreAndPostConditions` ? :-)